pages/rfc/registry-updates.rst

```diff
@@ -0,0 +1,94 @@
+.. _nuts-documentation-registry-updates:
+
+RFC: Registry Updates
+#####################
+
+Motivation
+**********
+
+The Nuts Registry contains information about who (vendors, organizations) participates in the Nuts network and where
+to find the resources they're exposing through endpoint registration. Parties also use it to register their certificates
+so that other parties can verify the integrity and authenticity of exposed data.
+
+The registry is modelled as an append-only event log which makes it easy to add new data (e.g. register an endpoint).
+However, there is currently no way to alter an data (e.g. rename an organization). This is an essential feature for any
+Nuts network which is managed by more than a single party, since altering data currently requires coordinated data
+pruning across all participating Nuts nodes.
+
+This document proposes a mechanism for updating data in the registry.
+
+Status of this RFC
+******************
+
+DRAFT
+
+Terminology
+***********
+
+- Event: an event in the registry event log, e.g. the RegisterVendor event
+- Object: the thing an event describes; either a vendor, organization or endpoint.
+- Event header: top-level properties of an event describing its payload.
+
+Mechanics
+*********
+
+Since events itself are immutable, only new events can be added to describe an object's new state. A subsequent event
+completely replaces the object's state described by the preceding event. The subsequent event must refer to the event
+it revises by its hash.
+
+Event referral
+==============
+
+An event refers to a preceding event (which it replaces) by its hash. There can only be one event referring to a particular
```

Comment:
This would require consistency/ordering and thus some sort of consensus algorithm...

Reply:
Could it be done without consensus and with branching?

Reply:
Since every object (vendor/organization/endpoint) is owned and managed by a single party (vendor), I'd make that party responsible for guaranteeing the events they emit aren't clashing.

A clash could be 2 events referring to the same preceding event. We could add a timestamp to the payload as to allow a receiver to choose the youngest event. The emitting party could spoof the timestamp but it would 'shoot itself in the foot' by doing so, since it just corrupts its own data.

I'm not sure how branching/merging would work but wouldn't that complicate things since the data is managed by a single party, and thus responsible for correcting errors that come from 'double-spending'-esque situations?

Reply:
'shoot itself in the foot' in this case would be should all nodes in the foot.

Maybe the best approach would be to use a CRDT:

- use a sequence number (generated by clustered DB) concatenated by a timestamp (or use timestamp from the sig). lowest number wins.
- multiple entries with the same payload hash are the same. 
- multiple entries with the same seq.number that refer to the same entry both survive and it's data is both valid (with different hashes, should be extremely rare)
- updates on multiple entries refer to one of the entries and longest branch wins OR updates on multiple entries refer to the merkle value of the two previous entries.

This should have all the nodes determine the same state even when a node 'shoots itself in the foot'

Reply:
So if due to timing issues of internal vendor events (large enterprise setting) 2 updates reach a Nuts node at the same time and there's a split brain in the enterprise DB causing it to generate the same seq number for both nodes (bad DB!), a race condition could occur causing some Nuts nodes in the network to see event1 first and others event2. Above mentioned rules would then approve both. If this case would, for example, occur in publishing a new pub.key. Two keys will be active but only 1 will be accepted by the faulty node. This will generate errors at the faulty node, the administrator will have to inspect the settings and see a double entry and fix it by deleting one.